File: Seed/Scripts/heartbeat_gate.py
#!/usr/bin/env python3
"""Lightweight activity gate for the workspace heartbeat.

This script does not decide what to change. It gives the heartbeat a cheap,
structured snapshot so the agent can avoid deep review when nothing happened.
"""

from __future__ import annotations

import argparse
import json
import subprocess
from datetime import datetime, timezone
from pathlib import Path


def read_text(path: Path) -> str:
    try:
        return path.read_text(encoding="utf-8")
    except FileNotFoundError:
        return ""


def has_pending_markdown_items(path: Path) -> bool:
    text = read_text(path)
    if not text.strip():
        return False
    for line in text.splitlines():
        stripped = line.strip()
        if stripped.startswith("- ") and not stripped.lower().startswith("- no "):
            return True
    return False


def status_path_mtime(workspace: Path, status_line: str) -> float:
    raw_path = status_line[3:].strip()
    if " -> " in raw_path:
        raw_path = raw_path.split(" -> ", 1)[1].strip()
    raw_path = raw_path.strip('"')
    path = workspace / raw_path
    try:
        return path.stat().st_mtime
    except OSError:
        # Deleted, renamed, or shell-escaped paths still represent activity.
        # Keep them visible instead of filtering them out as stale.
        return float("inf")


def git_status(workspace: Path, since: float) -> list[str]:
    try:
        result = subprocess.run(
            ["git", "status", "--short"],
            cwd=workspace,
            check=False,
            text=True,
            capture_output=True,
        )
    except FileNotFoundError:
        return ["git-unavailable"]
    if result.returncode != 0:
        return [f"git-status-failed: {result.stderr.strip()}"]
    ignored_prefixes = (
        "?? .business-ai-kit/source/",
        "!! .business-ai-kit/source/",
    )
    lines = []
    for line in result.stdout.splitlines():
        if not line.strip() or line.startswith(ignored_prefixes):
            continue
        if since and status_path_mtime(workspace, line) <= since:
            continue
        lines.append(line)
    return lines


def newest_mtime(paths: list[Path]) -> float:
    mtimes = []
    for path in paths:
        if path.exists():
            mtimes.append(path.stat().st_mtime)
    return max(mtimes) if mtimes else 0.0


def recent_session_candidates(workspace: Path, since: float) -> list[str]:
    sessions_root = Path.home() / ".codex" / "sessions"
    if not sessions_root.exists():
        return []

    candidates: list[str] = []
    workspace_name = workspace.name
    workspace_text = str(workspace)
    for path in sessions_root.rglob("*"):
        if not path.is_file():
            continue
        try:
            if path.stat().st_mtime <= since:
                continue
            sample = path.read_text(encoding="utf-8", errors="ignore")[:200_000]
        except OSError:
            continue
        if workspace_text in sample or workspace_name in sample:
            candidates.append(str(path))
        if len(candidates) >= 20:
            break
    return candidates


def iso(ts: float) -> str | None:
    if not ts:
        return None
    return datetime.fromtimestamp(ts, tz=timezone.utc).isoformat()


def main() -> int:
    parser = argparse.ArgumentParser(description="Check whether heartbeat work is likely needed.")
    parser.add_argument("--workspace", default=".", help="Workspace path. Defaults to current directory.")
    args = parser.parse_args()

    workspace = Path(args.workspace).resolve()
    instructions = workspace / "Agent-Instructions"
    automation_log = instructions / "Automation-Log.md"
    outbox = instructions / "Outbox.md"
    last_state_time = newest_mtime([automation_log, outbox])

    inbox_pending = has_pending_markdown_items(instructions / "Inbox.md")
    incoming_signal = has_pending_markdown_items(instructions / "Signals" / "Incoming.md")
    outgoing_signal = has_pending_markdown_items(instructions / "Signals" / "Outgoing.md")
    status_lines = git_status(workspace, last_state_time)
    session_candidates = recent_session_candidates(workspace, last_state_time)

    reasons = []
    if inbox_pending:
        reasons.append("pending inbox items")
    if incoming_signal or outgoing_signal:
        reasons.append("pending signals")
    if status_lines:
        reasons.append("workspace git changes")
    if session_candidates:
        reasons.append("recent Codex session evidence")

    result = {
        "workspace": str(workspace),
        "checked_at": datetime.now(tz=timezone.utc).isoformat(),
        "last_state_time": iso(last_state_time),
        "needs_deep_review": bool(reasons),
        "reasons": reasons,
        "signals": {
            "inbox_pending": inbox_pending,
            "incoming_signal": incoming_signal,
            "outgoing_signal": outgoing_signal,
            "git_status_count": len(status_lines),
            "recent_session_candidate_count": len(session_candidates),
        },
        "git_status": status_lines[:50],
        "recent_session_candidates": session_candidates,
    }
    print(json.dumps(result, indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: Seed/Scripts/heartbeat_gate.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import { spawnSync } from 'node:child_process';

function parseArgs(argv) {
  const args = { workspace: '.' };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === '--workspace') {
      args.workspace = argv[index + 1] || '.';
      index += 1;
    } else if (arg === '-h' || arg === '--help') {
      printHelp();
      process.exit(0);
    }
  }
  return args;
}

function printHelp() {
  console.log(`Usage: node Scripts/heartbeat_gate.mjs [--workspace .]

Check whether heartbeat work is likely needed.`);
}

function readText(filePath) {
  try {
    return fs.readFileSync(filePath, 'utf8');
  } catch {
    return '';
  }
}

function hasPendingMarkdownItems(filePath) {
  const text = readText(filePath);
  if (!text.trim()) return false;
  return text.split(/\r?\n/).some((line) => {
    const stripped = line.trim();
    return stripped.startsWith('- ')
      && !stripped.toLowerCase().startsWith('- no ')
      && !/^- \[x\]/i.test(stripped);
  });
}

function fileMtime(filePath) {
  try {
    return fs.statSync(filePath).mtimeMs / 1000;
  } catch {
    return 0;
  }
}

function gitStatus(workspace) {
  const result = spawnSync('git', ['status', '--short'], {
    cwd: workspace,
    encoding: 'utf8',
  });
  if (result.error && result.error.code === 'ENOENT') return ['git-unavailable'];
  if (result.status !== 0) return [`git-status-failed: ${(result.stderr || '').trim()}`];

  const ignoredPrefixes = ['?? .business-ai-kit/source/', '!! .business-ai-kit/source/'];
  const lines = [];
  for (const line of result.stdout.split(/\r?\n/)) {
    if (!line.trim()) continue;
    if (ignoredPrefixes.some((prefix) => line.startsWith(prefix))) continue;
    lines.push(line);
  }
  return lines;
}

function newestMtime(paths) {
  const mtimes = [];
  for (const filePath of paths) {
    try {
      mtimes.push(fs.statSync(filePath).mtimeMs / 1000);
    } catch {
      // Missing state files are treated as no prior heartbeat state.
    }
  }
  return mtimes.length ? Math.max(...mtimes) : 0;
}

function iso(timestampSeconds) {
  if (!timestampSeconds) return null;
  return new Date(timestampSeconds * 1000).toISOString();
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const workspace = path.resolve(args.workspace);
  const instructions = path.join(workspace, 'Agent-Instructions');
  const automationLog = path.join(instructions, 'Automation-Log.md');
  const outbox = path.join(instructions, 'Outbox.md');
  const lastStateTime = newestMtime([automationLog, outbox]);

  const inboxPath = path.join(instructions, 'Inbox.md');
  const incomingSignalPath = path.join(instructions, 'Signals', 'Incoming.md');
  const outgoingSignalPath = path.join(instructions, 'Signals', 'Outgoing.md');
  const inboxPending = hasPendingMarkdownItems(inboxPath);
  const incomingSignal = hasPendingMarkdownItems(incomingSignalPath);
  const outgoingSignal = hasPendingMarkdownItems(outgoingSignalPath);
  const inboxChanged = inboxPending && fileMtime(inboxPath) > lastStateTime;
  const signalsChanged =
    (incomingSignal && fileMtime(incomingSignalPath) > lastStateTime) ||
    (outgoingSignal && fileMtime(outgoingSignalPath) > lastStateTime);
  const statusLines = gitStatus(workspace);

  const reasons = [];
  if (inboxChanged) reasons.push('changed inbox items');
  if (signalsChanged) reasons.push('changed signals');
  if (statusLines.length) reasons.push('workspace git changes');

  const result = {
    workspace,
    checked_at: new Date().toISOString(),
    last_state_time: iso(lastStateTime),
    needs_deep_review: Boolean(reasons.length),
    reasons,
    signals: {
      inbox_pending: inboxPending,
      inbox_changed_since_last_state: inboxChanged,
      incoming_signal: incomingSignal,
      outgoing_signal: outgoingSignal,
      signals_changed_since_last_state: signalsChanged,
      git_status_count: statusLines.length,
    },
    git_status: statusLines.slice(0, 50),
    // This is a local hint, not a scheduler or a complete activity history.
    scope: 'workspace files only; no assistant session history is read',
    limitations: 'Activity hint only. Log timestamps are not processing checkpoints; clean commits and deadlines require an explicit scoped check.',
  };

  console.log(JSON.stringify(result, null, 2));
  return 0;
}

process.exit(main());
